ships/0037-shipwright-build-strategy-controller.md

```diff
@@ -0,0 +1,190 @@
+In order to improve user experience around Shipwright, the operator should also install the build strategies
+
+### Goals
+
+- Create a Build Strategy catalog repository in the upstream shipwright-io organization
+- Copy/publish the Build strategies from the upstream build repository to the new create catalog repository
+- Allow a user to install the default Build Strategy Catalog or if applicable the user provided catalog, via the Shipwright Operator
+ 
+### Non-Goals
+
+- Add a new operator to install the Build Strategy catalog
+- Curate or rework the Build Strategies in the upstream build repository when copied to the new catalog repository
+
+## Proposal
+
+We propose to introduce a new CRD in the Shipwright Operator : `ShipwrightBuildStrategy` API.
+
+```yaml
+apiVersion: operator.shipwright.io/v1alpha1
+kind: ShipwrightBuildStrategy
+metadata:
+  name: buildStrategy
+spec:
+  targetNamespace: shipwright-build
+  catalog:
+    url: https://github.com/shipwright-io/buildstrategy-catalog
+```
+We propose to add a new controller to the Shipwright Operator that reconciles the ShipwrightBuildStrategy.
+
+
+### User Stories [optional]
+
+#### Story 1
+
+As a Kubernetes cluster administrator
+I want to be able to install the default catalog of Shipwright Builds Strategies via the Shipwright Operator
+So that I can provide Shipwright Build Strategies to my developers
+
+
+#### Story 2
+
+As a Kubernetes cluster administrator
+I want to be able to install , with an url I provide, my own catalog of Shipwright Builds Strategies via the Shipwright Operator
+So that I can provide a custom subset of Shipwright Build Strategies to my developers
+
+### Implementation Notes
+
+#### API
+
+We propose to add a new custom resource  the `ShipwrightBuildStrategy`.
+
+```yaml
+apiVersion: operator.shipwright.io/v1alpha1
+kind: ShipwrightBuildStrategy
+metadata:
+  name: buildStrategy
+spec:
+  targetNamespace:
+  catalog:
+    url:
+```
+
+#### Controller
+
+Actually the Shipwright Operator uses controller-runtime package to bootstrap a manager 
+and setup the Shipwright Build controller within that manager.
+In the target implementation we will continue using controller-runtime
+We will split the controllers each acheiving a specific function in its own package:
+- shipwrightbuild controller to reconcile ShipwrightBuild CRD and installs Shipwright Build manifests
+- shipwrightbuilstrategy controller to reconcile  ShipwrightBuildStrategy CRD and installs Shipwright Build Strategy catalog.
+The main package will bootstrap a controller manager and adds all the controllers to the manager.
+We propose this layer for package organisation
+
+```go
+main.go 
+controllers
+  controller.go
+  add_shipwrightbuild.go
+  add_shipwrightbuildstrategy.go
+  shipwrightbuild
+    shipwrightbuild_controller.go
+  shipwrightbuildstragegy
+     shipwrightbuildstrategy_controller.go
+```
```

Comment:
This feels like deeper implementation details - we don't want to be overly prescriptive in the proposal when it comes to how this functionality is delivered.

Reply:
You are right, I will update the SHIP and add only requireemnt or high level design recommandation
We do want One single manager and multiple controllers assigned  to it, which means e require a new controller for installing the strategies